### PR TITLE
[#8918] Instructor: view results: questions for ranking: show rank within team

### LINKED ISSUE #8918: Instructor: view results: questions for ranking: show rank within team
Context: Instructor viewing results of a session that has a question for ranking items (e.g., rank team members based on technical competence)

Current: In question stats, we show overall ranking of the students/items based on all rankings received

Suggested: In addition, also show ranking within the team when that makes sense e.g., Amy, Ben, Charlie is ranked 3, 23, 56 within the course but ranked 1, 2, 3 within the team.

Justification: Some instructors might base their grade on the rank within the team. Currently, they have to do an extra step to figure out the rank within the team.

Notes: the stat can be included in the downloaded csv as well

### PR DESCRIPTION
**Pull request by AdithyaNarayan**
Fixes #8918 

**Outline of Solution**
- Created new member fields in the `rank-recipient-question-statistics-calculation` class for team rank.
  - In `calculateRanks`, all the teams and their team members are captured in the `optionsPerTeam` variable, which maps team names to its team members.
  - For each team, the `calculateRankPerOption` is called with only the team members' responses, and the ranks within each team are calculated.
- Likewise, the process is repeated for team rank excluding self.
- Added test case for ranks within a team.

**Screenshot**
pic1
pic2

**Points to Note**
The solution assumes that team names are unique.

### CODE REVIEW

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Thanks for the PR 👍 

We should only be showing the team rank when it makes sense (we should check against the `recipientType`) 

For example, when ranking teams there is no need to show team rank:

Screenshot 2020-12-31 at 6 42 45 PM

**Comment by AdithyaNarayan** [comment]
Okay, thanks for the comment! I have made sure that the team ranking is shown only when the  `recipientType` is `OWN_TEAM_MEMBERS` or `OWN_TEAM_MEMBERS_INCLUDING_SELF`.

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/rank-recipients-question-statistics-calculation.ts
Comment (rrtheonlyone) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/rank-recipients-question-statistics-calculation.ts:70]: the naming here is a little confusing - we are referring to `recipient` but the variable names all somehow say `option`. 

I think we might be able to streamline this entire statistic calculation with a combination of `map`, `filter` and `reduce` (in a functional manner without for loops).  However, this I leave it up to you: as long as it's readable I am okay.

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments on the implementation/tests 

Also do update the help section as well to explain what the team rank is all about

**Comment by AdithyaNarayan** [comment]
Done👍

**Comment by AdithyaNarayan** [comment]
Is there anything more to do for this PR?

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯 Nice one

Apologies for delay in this @AdithyaNarayan .. should be good to go once CI passes

File: src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts
Comment (madanalogy) [inline src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts:11]: Shouldn't this be indented? But okay not a big deal since the lint check passes

File: src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html
Comment (madanalogy) [inline src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html:662]: Brackets not needed

File: src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts
Comment (madanalogy) [inline src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts:12]: Any particular reason why test data is stored in a separate file? I'm not entirely against the approach don't get me wrong, just that our frontend tests so far all have their data embedded in the `spec.ts` file so I'm wondering if you had any particular reason for the extraction of data. 

I can understand the extraction if the same data is being used by other test components but this doesn't seem to be the case. Would be better for now to just have the data as a constant variable as in other frontend tests for consistency (e.g. `instructor-home-page.component.spec.ts`)

File: src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html
Comment (madanalogy) [inline src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html:679]: Typo in the sentence, I'm assuming the period at the end of "team" was unintended

File: src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html
Comment (madanalogy) [inline src/web/app/pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-section.component.html:681]: Unnecessary starting `And`, can be removed

**Review by madanalogy — CHANGES_REQUESTED** [review]

**Comment by AdithyaNarayan** [comment]
Addressed the comments. Thanks for the feedback

File: src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts
Comment (AdithyaNarayan) [reply src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts]: Just felt that the code looked bad cosmetically with sample data and decided to put it in a JSON. Reverted

Comment (madanalogy) [reply src/web/app/components/question-types/question-statistics/rank-recipients-question-statistics.component.spec.ts]: You could consider putting up an issue for this actually, I remember it was something I wanted to raise but didn't have the time to. That is to say, refactor frontend test data into a dedicated resource folder for reusability instead of having it nested within the `spec.ts` files. We can have a community discussion about this, and if the idea is adopted start taking the approach as an epic/task.

**Review by madanalogy — APPROVED** [review]
LGTM! Will merge assuming no issues with the CI checks

**Pull request merged**
